internal/controllers/topology/cluster/desired_state.go

```diff
@@ -427,10 +447,50 @@ func computeCluster(_ context.Context, s *scope.Scope, infrastructureCluster, co
 
 	// Set the references to the infrastructureCluster and controlPlane objects.
 	// NOTE: Once set for the first time, the references are not expected to change.
-	cluster.Spec.InfrastructureRef = contract.ObjToRef(infrastructureCluster)
-	cluster.Spec.ControlPlaneRef = contract.ObjToRef(controlPlane)
+	var err error
+	cluster.Spec.InfrastructureRef, err = calculateDesiredRef(cluster.Spec.InfrastructureRef, infrastructureCluster)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to calculate infrastructureRef")
+	}
+	cluster.Spec.ControlPlaneRef, err = calculateDesiredRef(cluster.Spec.ControlPlaneRef, controlPlane)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to calculate controlPlaneRef")
+	}
+
+	return cluster, nil
+}
 
-	return cluster
+// calculateDesiredRef calculates the desired ref based on a currentRef and a desiredReferencedObject.
+// The goal is to calculate a ref which points to desiredReferencedObject while not overwriting version
+// bumps which could have been done by another controller.
+// If currentRef is not set, it just returns the ref to the desiredReferencedObject
+// If currentRef is set and the group is different from the one in desiredReferencedObject, it returns
+// the ref to the desiredReferencedObject.
+// If currentRef is set and the group is the same as the one in desiredReferencedObject, it returns
+// the ref to the desiredReferencedObject but with the apiVersion from currentRef to not overwrite
+// the version bump which could have been done by another controller.
```

Comment:
```suggestion
```suggestion
// calculateDesiredRef updates the desired ref so it doesn't override the APIversion stored in the current ref, if any; this is required because the APIVersion in the desired ref is aligned to the APIVersion used in Cluster Class when reading the current state.
If the current ref is nil or GK doesn't match, no changes are applied to the desired ref.
```
```